### PR TITLE
Initial support for ServerReflection in the backend proxy

### PR DESCRIPTION
**Pull request by viktorklang**
TODO: Write automated tests for this feature, this has been mostly tested using the grpc_cli tool.

Sample output can be seen here:

```
─○ grpc_cli ls localhost:9000
com.example.shoppingcart.ShoppingCart

─○ grpc_cli ls localhost:9000 com.example.shoppingcart.ShoppingCart -l
filename: /Users/viktorklang/Documents/workspace/stateful-serverless/src/samples/js-shopping-cart/proto/shoppingcart.proto
package: com.example.shoppingcart;
service ShoppingCart {
  rpc AddItem(com.example.shoppingcart.AddLineItem) returns (com.example.shoppingcart.Empty) {}
  rpc RemoveItem(com.example.shoppingcart.RemoveLineItem) returns (com.example.shoppingcart.Empty) {}
  rpc GetCart(com.example.shoppingcart.GetShoppingCart) returns (com.example.shoppingcart.Empty) {}
}

─○ grpc_cli type localhost:9000 com.example.shoppingcart.AddLineItem -l
message AddLineItem {
  optional string userId = 1 [default = "", (.lightbend.serverless.entity_key) = true];
}

─○ grpc_cli type localhost:9000 com.example.shoppingcart.AddLineI -l
Type com.example.shoppingcart.AddLineI not found.
```

### CODE REVIEW

File: src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala [hunk 1]
```diff
@@ -179,3 +176,114 @@ object Serve {
...
+    implicit val ec: ExecutionContext = mat.executionContext
+    import ServerReflection.Serializers._
+
+    val handler = handle(fileDesc)
+
+    {
+      case req: HttpRequest if req.uri.path == ReflectionPath =>
+        val responseCodec = Codecs.negotiate(req)
+        GrpcMarshalling.unmarshalStream(req)(ServerReflectionRequestSerializer, mat)
+        .map( _.alsoTo(Sink.foreach(x => println(x.toString))).via(handler))
```
Comment (raboof) [inline src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala]: println

File: src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala [hunk 2]
```diff
@@ -179,3 +176,114 @@ object Serve {
...
+      fileDesc.getExtensions.iterator.asScala.collect({ case ext if ext.getFullName == container => ext.getNumber }).toList
+    )((list, fd) => findExtensionNumbersForContainingType(container, fd) ::: list)
+
+  // TODO create caches for all of these lookups when/if needed
+  private def handle(fileDesc: FileDescriptor): Flow[ServerReflectionRequest, ServerReflectionResponse, NotUsed] =
+    Flow[ServerReflectionRequest].map { req =>
+      import ServerReflectionRequest.{ MessageRequest => In}
+      import ServerReflectionResponse.{ MessageResponse => Out}
+
+      val response = req.messageRequest match {
```
Comment (raboof) [inline src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala]: Looks useful, might indeed be interesting to make this easy to use with akka-grpc in some form.

File: src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala [hunk 3]
```diff
@@ -179,3 +176,114 @@ object Serve {
...
+          Out.Empty
+        case In.FileByFilename(fileName) =>
+          val list = findFileDescForName(fileName, fileDesc) match {
+            case None => Nil // throw new GrpcServiceException(Status.NOT_FOUND.augmentDescription(s"File not found: $fileName"))
+            case Some(file) => file.toProto.toByteString :: Nil
+          }
+          Out.FileDescriptorResponse(FileDescriptorResponse(list))
+        case In.FileContainingSymbol(symbol) =>
+          val list = findFileDescForSymbol(symbol, fileDesc) match {
+            case None => Nil // throw new GrpcServiceException(Status.NOT_FOUND.augmentDescription(s"Symbol not found: $symbol"))
```
Comment (raboof) [inline src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala]: why not?

Comment (viktorklang) [reply src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala]: There is no spec. And the client (esp. grpc_cli) will show some pretty horrible error messages for lookups it decides to perform on its own. :S

File: src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala (hunk 2 above)
Comment (viktorklang) [reply src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala]: :+1:

File: src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala (hunk 1 above)
Comment (viktorklang) [reply src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala]: Doh! Good spotted, that's naturally there for me trying to reverse-engineer the spec.

**Comment by viktorklang** [comment]
@jroper Supporting this feature means updating the node.js implementation to send all the correct types:   rpc GetCart(com.example.shoppingcart.GetShoppingCart) returns (com.example.shoppingcart.Empty) {} <--- it doesn't return Empty really.

**Pull request merged**